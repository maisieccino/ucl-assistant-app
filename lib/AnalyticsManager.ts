import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

import configureStore from "../configureStore"

const { store } = configureStore

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = (): Promise<void> => Amplitude.initialize(AMPLITUDE_API_KEY)

const shouldTrackAnalytics = (): boolean => (
  AMPLITUDE_API_KEY
  && !__DEV__
  && store.getState().user.settings.shouldTrackAnalytics
)

const setUserId = (userId): void => {
  if (!shouldTrackAnalytics()) { return }

  // make sure this is unique
  if (userId) {
    Amplitude.setUserId(userId)
  }
}

const setUserProperties = ({ ...userProperties }): void => {
  if (!shouldTrackAnalytics()) { return }

  Amplitude.setUserProperties({ ...userProperties })
}

const clearUserProperties = (): void => {
  Amplitude.clearUserProperties()
}

const logEvent = (eventName: string, eventProperties?): void => {
  if (!shouldTrackAnalytics()) { return }

  if (eventProperties) {
    Amplitude.logEventWithProperties(eventName, eventProperties)
  } else {
    Amplitude.logEvent(eventName)
  }
}

const logScreenView = (screenName: string): void => {
  if (!shouldTrackAnalytics()) { return }

  logEvent(`VIEW_SCREEN_${screenName}`)
}

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

export const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default AnalyticsManager
