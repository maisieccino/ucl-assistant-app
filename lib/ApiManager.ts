import axios from "axios"
import { encode as btoa } from "base-64"

import {
  EQUIPMENT_URL,
  FREE_ROOMS_URL,
  ROOMBOOKINGS_URL,
  ROOMS_URL,
  SITES_URL,
  TIMETABLE_URL,
  WORKSPACES_URL,
} from "../constants/API"
import ErrorManager from "./ErrorManager"
import LocalisationManager from './LocalisationManager'

const roomsController = {
  getBookings: async (
    token = null,
    {
      roomid,
      siteid,
      date = LocalisationManager.getMoment().format(`YYYYMMDD`),
    },
  ) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          date,
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.bookings
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getEmptyRooms: async (
    token = null,
    {
      startDateTime = (new Date()).toISOString(),
      endDateTime = (
        LocalisationManager
          .getMoment()
          .add(1, `hours`)
          .toISOString()
      ),
    } = {},
  ) => {
    try {
      const results = await axios.get(FREE_ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          endDateTime,
          startDateTime,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.free_rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getEquipment: async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.equipment
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getSites: async (token = null) => {
    try {
      const results = await axios.get(SITES_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.sites
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  search: async (token = null, query) => {
    if (query && query.length < 3) {
      return {}
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          query,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
}

const workspacesApi = axios.create({
  baseURL: WORKSPACES_URL,
})

const workspacesController = {
  getLiveImage: async (token = null, { surveyId, mapId }) => {
    try {
      const result = await workspacesApi.get(`/getliveimage/map.svg`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          map_id: mapId,
          survey_id: surveyId,
        },
        responseType: `arraybuffer`,
      })
      const image = btoa(
        new Uint8Array(result.data).reduce(
          (data, byte) => data + String.fromCharCode(byte),
          ``,
        ),
      )
      // there may be no need to base64 SVGs
      // but I have yet to test this extensively
      // https://css-tricks.com/probably-dont-base64-svg/
      return `data:image/svg+xml;base64,${image}`
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getSummary: async (token = null) => {
    try {
      const { data, headers } = await workspacesApi.get(`/summary`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!data.content || data.error !== ``) {
        throw new Error(data.error || `There was a problem`)
      }
      return {
        data: data.content,
        lastModified: LocalisationManager.parseToMoment(
          headers[`last-modified`],
        ),
      }
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getWorkspaces: async (token = null) => {
    try {
      const { data, headers } = await workspacesApi.get(`/`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!data.content || data.error !== ``) {
        throw new Error(data.error)
      }
      return {
        data: data.content,
        lastModified: LocalisationManager.parseToMoment(
          headers[`last-modified`],
        ),
      }
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
}

const timetableApi = axios.create({
  baseURL: TIMETABLE_URL,
})

const timetableController = {
  getPersonalTimetable: async (
    token = null,
    date = LocalisationManager.getMoment(),
  ) => {
    const datePart = date
      ? `?date=${date.clone().startOf(`isoWeek`).format(`YYYY-MM-DD`)}`
      : ``
    const headers = {
      authorization: `Bearer ${token}`,
    }
    ErrorManager.addDetail({ headers })
    const results = await timetableApi.get(datePart, { headers })
    if (!results.data.content.ok || results.data.error !== ``) {
      throw new Error(
        results.data.error
        || `There was a problem fetching your personal timetable`,
      )
    }
    const { data: { content: { timetable } } } = results
    Object.keys(timetable).forEach((day) => {
      const dayTimetable = timetable[day]
      timetable[day] = {
        lastModified: LocalisationManager.parseToMoment(
          results.headers[`last-modified`],
        ),
        timetable: dayTimetable,
      }
    })
    return timetable
  },
  getPersonalWeekTimetable: async (
    token = null,
    date = LocalisationManager.getMoment().startOf(`isoWeek`),
  ) => {
    const queryParam = `?date=${date.format(`YYYY-MM-DD`)}`
    const headers = {
      authorization: `Bearer ${token}`,
    }
    ErrorManager.addDetail({ headers })
    const results = await timetableApi.get(`/week/${queryParam}`, {
      headers,
    })
    if (results.data.error !== ``) {
      throw new Error(
        results.data.error
        || `An error occurred when fetching your personal weekly timetable`,
      )
    }
    const { data: { content: { timetable } } } = results
    // eslint-disable-next-line sonarjs/no-identical-functions
    Object.keys(timetable).forEach((day) => {
      const dayTimetable = timetable[day]
      timetable[day] = {
        lastModified: LocalisationManager.parseToMoment(
          results.headers[`last-modified`],
        ),
        timetable: dayTimetable,
      }
    })
    return timetable
  },
}

export default {
  rooms: roomsController,
  timetable: timetableController,
  workspaces: workspacesController,
}
