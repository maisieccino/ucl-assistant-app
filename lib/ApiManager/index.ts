import PeopleController from './PeopleController'
import RoomsController from './RoomsController'
import TimetableController from './TimetableController'
import WorkspacesController from './WorkspacesController'

export default {
  people: PeopleController,
  rooms: RoomsController,
  timetable: TimetableController,
  workspaces: WorkspacesController,
}
