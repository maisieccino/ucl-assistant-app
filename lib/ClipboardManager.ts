import { Clipboard } from 'react-native'

const {
  getString,
  setString,
} = Clipboard

export default {
  getString,
  setString,
}
