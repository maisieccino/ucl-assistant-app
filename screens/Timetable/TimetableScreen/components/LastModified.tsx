import moment from 'moment'
import React, { useMemo } from 'react'
import { StyleSheet, View } from 'react-native'
import { CentredText, ErrorText, Link } from "../../../../components/Typography"
import { LocalisationManager } from "../../../../lib"

const styles = StyleSheet.create({
  error: {
    marginBottom: 20,
  },
  lastModified: {
    alignItems: `center`,
    width: `100%`,
  },
})

interface Props {
  date?: string,
  isLoading?: boolean,
  lastModified: moment.Moment,
  openFAQ?: () => void,
}

const LastModified:React.FC<Props> = ({
  lastModified,
  openFAQ = (): void => {},
  isLoading = true,
  date = LocalisationManager.getMoment().toISOString(),
}) => {
  const isStale = useMemo(() => (
    !isLoading
    && lastModified
    && lastModified.isBefore(
      LocalisationManager.getMoment().subtract(24, `hour`),
    )
    // not stale if the data is more recent than the date
    && !lastModified.isAfter(
      LocalisationManager.parseToMoment(date).endOf(`isoWeek`),
    )
  ), [isLoading, lastModified, date])

  const lastUpdatedText = useMemo(() => (
    lastModified
    && lastModified.isBefore()
      ? lastModified.fromNow().toLowerCase()
      : `just now`
  ), [lastModified])

  return (
    <View style={styles.lastModified}>
      {isStale ? (
        <ErrorText containerStyle={styles.error}>
          Our timetable data is stale. Sorry about that.
          We&apos;re working on getting this fixed as quickly as possible.
        </ErrorText>
      ) : null}
      <Link onPress={openFAQ}>
        <CentredText>
          {`Last updated ${lastUpdatedText}`}
        </CentredText>
      </Link>
    </View>
  )
}

export default LastModified
