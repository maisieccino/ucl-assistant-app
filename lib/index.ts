import AnalyticsManager from './AnalyticsManager'
import ApiManager from './ApiManager'
import AssetManager from './AssetManager'
import debounce from './debounce'
import DeviceManager from './DeviceManager'
import ErrorManager from './ErrorManager'
import LocalisationManager from './LocalisationManager'
import MailManager from './MailManager'
import MapsManager from './MapsManager'
import PushNotificationsManager from './PushNotificationsManager'
import Shadow from './Shadow'
import WebBrowserManager from './WebBrowserManager'

const Random = {
  array: (arr) => arr[Math.floor(Math.random() * arr.length)],
}

export {
  AnalyticsManager,
  ApiManager,
  AssetManager,
  DeviceManager,
  ErrorManager,
  LocalisationManager,
  MailManager,
  MapsManager,
  PushNotificationsManager,
  Random,
  Shadow,
  WebBrowserManager,
  debounce,
}
