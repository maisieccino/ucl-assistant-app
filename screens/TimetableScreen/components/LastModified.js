import PropTypes from "prop-types"
import React from 'react'
import { momentObj } from "react-moment-proptypes"
import { StyleSheet, View } from 'react-native'

import { CentredText, ErrorText, Link } from "../../../components/Typography"
import { LocalisationManager } from "../../../lib"

const styles = StyleSheet.create({
  error: {
    marginBottom: 20,
  },
  lastModified: {
    alignItems: `center`,
    width: `100%`,
  },
})

class LastModified extends React.Component {
  static propTypes = {
    date: PropTypes.string,
    isLoading: PropTypes.bool,
    lastModified: PropTypes.oneOfType([momentObj, PropTypes.string]),
    openFAQ: PropTypes.func,
  }

  static defaultProps = {
    date: LocalisationManager.getMoment(),
    isLoading: true,
    lastModified: null,
    openFAQ: () => { },
  }

  renderError = () => (
    <ErrorText containerStyle={styles.error}>
      Our timetable data is stale. Sorry about that.
      We&apos;re working on getting this fixed as quickly as possible.
    </ErrorText>
  )

  render() {
    const {
      lastModified, openFAQ, isLoading, date,
    } = this.props

    if (lastModified === null || typeof lastModified !== `object`) {
      return null
    }

    const isStale = (
      !isLoading
      && lastModified.isBefore(
        LocalisationManager.getMoment().subtract(24, `hour`),
      )
      // not stale if the data is more recent than the date
      && !lastModified.isAfter(
        LocalisationManager.parseToMoment(date).endOf(`isoweek`),
      )
    )

    return (
      <View style={styles.lastModified}>
        {isStale ? this.renderError() : null}
        <Link onPress={openFAQ}>
          <CentredText>
            {`Last updated ${
              lastModified.isBefore()
                ? lastModified.fromNow().toLowerCase()
                : `just now`
            }`}
          </CentredText>
        </Link>
      </View>
    )
  }
}

export default LastModified
