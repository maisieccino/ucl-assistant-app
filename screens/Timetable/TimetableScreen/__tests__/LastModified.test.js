/**
 * @jest-environment jsdom
 */
import "react-native"

import MockDate from 'mockdate'
import React from 'react'
import { cleanup, render } from "react-native-testing-library"

import { LocalisationManager } from '../../../../lib'
import LastModified from
  "../components/LastModified"

describe(`LastUpdated`, () => {
  MockDate.set(`2019-11-18T08:47:21.000Z`)

  afterEach(() => {
    cleanup()
  })

  it(`renders the LastModified component`, () => {
    const mockProps = {
      lastModified: LocalisationManager.getMoment().subtract(11, `hours`),
    }
    const wrapper = render(<LastModified {...mockProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`does not render if date is missing`, () => {
    const mockProps = {
      lastModified: undefined,
    }
    const wrapper = render(<LastModified {...mockProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`shows error when studyspace info is stale`, async () => {
    const mockProps = {
      lastModified: LocalisationManager
        .getMoment()
        .subtract(25, `hours`),
    }
    const wrapper = render(<LastModified {...mockProps} isLoading={false} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`does not show error when data is more recent than date`, async () => {
    const mockDate = `2019-09-08`
    const mockProps = {
      date: mockDate,
      lastModified: LocalisationManager
        .parseToMoment(mockDate)
        .add(24, `hours`),
    }
    const wrapper = render(<LastModified {...mockProps} isLoading={false} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
