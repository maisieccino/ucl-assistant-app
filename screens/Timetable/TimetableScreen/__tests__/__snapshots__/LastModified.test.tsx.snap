// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastUpdated does not render if date is missing 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "width": "100%",
          }
        }
      >
        <TouchableOpacity
          style={Object {}}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#4D5061",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 1,
                  "marginTop": 1,
                },
                Object {
                  "color": "#2980b9",
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#6E707F",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 5,
                    "marginTop": 5,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Last updated just now
            </Text>
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <TouchableOpacity
        style={Object {}}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#4D5061",
                "fontFamily": "apercu",
                "fontSize": 16,
                "marginBottom": 1,
                "marginTop": 1,
              },
              Object {
                "color": "#2980b9",
              },
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#6E707F",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 5,
                  "marginTop": 5,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Last updated just now
          </Text>
        </Text>
      </TouchableOpacity>
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LastUpdated does not show error when data is more recent than date 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "width": "100%",
          }
        }
      >
        <TouchableOpacity
          style={Object {}}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#4D5061",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 1,
                  "marginTop": 1,
                },
                Object {
                  "color": "#2980b9",
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#6E707F",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 5,
                    "marginTop": 5,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Last updated 2 months ago
            </Text>
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <TouchableOpacity
        style={Object {}}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#4D5061",
                "fontFamily": "apercu",
                "fontSize": 16,
                "marginBottom": 1,
                "marginTop": 1,
              },
              Object {
                "color": "#2980b9",
              },
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#6E707F",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 5,
                  "marginTop": 5,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Last updated 2 months ago
          </Text>
        </Text>
      </TouchableOpacity>
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LastUpdated renders the LastModified component 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "width": "100%",
          }
        }
      >
        <TouchableOpacity
          style={Object {}}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#4D5061",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 1,
                  "marginTop": 1,
                },
                Object {
                  "color": "#2980b9",
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#6E707F",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 5,
                    "marginTop": 5,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Last updated 11 hours ago
            </Text>
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <TouchableOpacity
        style={Object {}}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#4D5061",
                "fontFamily": "apercu",
                "fontSize": 16,
                "marginBottom": 1,
                "marginTop": 1,
              },
              Object {
                "color": "#2980b9",
              },
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#6E707F",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 5,
                  "marginTop": 5,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Last updated 11 hours ago
          </Text>
        </Text>
      </TouchableOpacity>
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LastUpdated shows error when studyspace info is stale 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "alignContent": "center",
                "alignItems": "center",
                "flexDirection": "row",
                "marginBottom": 20,
                "marginTop": 5,
              },
            ]
          }
        >
          <Text />
          <Text
            style={
              Array [
                Object {
                  "color": "rgba(215, 38, 61, 1)",
                  "flex": 1,
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 1,
                  "marginLeft": 5,
                  "marginTop": -1,
                },
                undefined,
              ]
            }
          >
            Our timetable data is stale. Sorry about that. We're working on getting this fixed as quickly as possible.
          </Text>
        </View>
        <TouchableOpacity
          style={Object {}}
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#4D5061",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 1,
                  "marginTop": 1,
                },
                Object {
                  "color": "#2980b9",
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#6E707F",
                    "fontFamily": "apercu",
                    "fontSize": 16,
                    "marginBottom": 5,
                    "marginTop": 5,
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Last updated a day ago
            </Text>
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "flexDirection": "row",
              "marginBottom": 20,
              "marginTop": 5,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            Array [
              Object {
                "color": "rgba(215, 38, 61, 1)",
                "flex": 1,
                "fontFamily": "apercu",
                "fontSize": 16,
                "marginBottom": 1,
                "marginLeft": 5,
                "marginTop": -1,
              },
              undefined,
            ]
          }
        >
          Our timetable data is stale. Sorry about that. We're working on getting this fixed as quickly as possible.
        </Text>
      </View>
      <TouchableOpacity
        style={Object {}}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#4D5061",
                "fontFamily": "apercu",
                "fontSize": 16,
                "marginBottom": 1,
                "marginTop": 1,
              },
              Object {
                "color": "#2980b9",
              },
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#6E707F",
                  "fontFamily": "apercu",
                  "fontSize": 16,
                  "marginBottom": 5,
                  "marginTop": 5,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Last updated a day ago
          </Text>
        </Text>
      </TouchableOpacity>
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
